contracts/modules/royalty/policies/IpRoyaltyVault.sol

```diff
@@ -0,0 +1,237 @@
+    /// @dev Only callable by the royalty policy LAP
+    function addIpRoyaltyVaultTokens(address token) external {
+        if (msg.sender != address(ROYALTY_POLICY_LAP)) revert Errors.IpRoyaltyVault__NotRoyaltyPolicyLAP();
+        _tokens.add(token);
+    }
+
+    /// @notice Snapshots the claimable revenue and royalty token amounts
+    /// @return snapshotId The snapshot id
+    function snapshot() external returns (uint256) {
+        if (block.timestamp - lastSnapshotTimestamp < ROYALTY_POLICY_LAP.getSnapshotInterval())
+            revert Errors.IpRoyaltyVault__SnapshotIntervalTooShort();
+
+        uint256 snapshotId = _snapshot();
+        lastSnapshotTimestamp = block.timestamp;
+
+        uint32 unclaimedTokens = unclaimedRoyaltyTokens;
+        unclaimedAtSnapshot[snapshotId] = unclaimedTokens;
+
+        address[] memory tokens = _tokens.values();
+
+        for (uint256 i = 0; i < tokens.length; i++) {
+            uint256 tokenBalance = IERC20Upgradeable(tokens[i]).balanceOf(address(this));
+            if (tokenBalance == 0) {
+                _tokens.remove(tokens[i]);
+                continue;
+            }
+
+            uint256 newRevenue = tokenBalance - claimVaultAmount[tokens[i]] - ancestorsVaultAmount[tokens[i]];
+            if (newRevenue == 0) continue;
+
+            uint256 ancestorsTokens = (newRevenue * unclaimedTokens) / totalSupply();
+            ancestorsVaultAmount[tokens[i]] += ancestorsTokens;
+
+            uint256 claimableTokens = newRevenue - ancestorsTokens;
+            claimableAtSnapshot[snapshotId][tokens[i]] = claimableTokens;
+            claimVaultAmount[tokens[i]] += claimableTokens;
+        }
+
+        emit SnapshotCompleted(snapshotId, block.timestamp, unclaimedTokens);
+
+        return snapshotId;
+    }
+
+    /// @notice Calculates the amount of revenue token claimable by a token holder at certain snapshot
+    /// @param account The address of the token holder
+    /// @param snapshotId The snapshot id
+    /// @param token The revenue token to claim
+    /// @return The amount of revenue token claimable
+    function claimableRevenue(address account, uint256 snapshotId, address token) external view returns (uint256) {
+        return _claimableRevenue(account, snapshotId, token);
+    }
+
+    /// @notice Allows token holders to claim revenue token based on the token balance at certain snapshot
+    /// @param snapshotId The snapshot id
+    /// @param tokens The list of revenue tokens to claim
+    function claimRevenueByTokenBatch(uint256 snapshotId, address[] calldata tokens) external nonReentrant {
+        for (uint256 i = 0; i < tokens.length; i++) {
+            uint256 claimableToken = _claimableRevenue(msg.sender, snapshotId, tokens[i]);
+            if (claimableToken == 0) continue;
+
+            isClaimedAtSnapshot[snapshotId][msg.sender][tokens[i]] = true;
+            claimVaultAmount[tokens[i]] -= claimableToken;
```

Comment:
Is there any risk claimVaultAmount[tokens[i]] < claimableToken?

Reply:
In that case, the code should revert as Solidity 0.8.x has an underflow guard for arithmetic operations ([see here](https://docs.soliditylang.org/en/latest/080-breaking-changes.html)).

We will need to fuzz-test the Royalty contracts. Theoretically the sum of `claimableToken` should never be more than `claimVaultAmount[tokens[i]]`.